gateway/engine/vertx-shareddata/src/main/java/io/apiman/gateway/engine/vertx/shareddata/SharedGlobalDataRegistry.java

```diff
@@ -0,0 +1,283 @@
+                    Future<Object> putNewApiKeyFuture = Future.future();
+                    Future<Object> endFuture = Future.future();
+
+                    // Order: Create new API Key reference; Replace old ID -> API mapping; Delete old key reference)
+                    // This should ensure no breaking/irreconcilable behaviour.
+                    objectMap.putIfAbsent(client.getApiKey(), client, putNewApiKeyFuture.completer());
+
+                    // Replace API Key reference
+                    putNewApiKeyFuture.compose(clientWithSameApiKey -> {
+                        Future<Object> replaceClientFuture = Future.future();
+                        // There's a small chance the same key will replace the old one, usually
+                        // only in hard-coded tests. Generally sameKeyReplace will be null.
+                        if (clientWithSameApiKey != null) {
+                            //System.err.println("!!!!! Same API Key -- Replacing. Must not delete later. !!!!!!");
+                            objectMap.replace(client.getApiKey(), client, replaceClientFuture.completer());
+                        } else {
+                            objectMap.putIfAbsent(clientIndex, client, replaceClientFuture.completer());
+                        }
+                        return replaceClientFuture;
+                    // Remove old API key reference
+                    }).compose(oldClientRaw -> {
+                        Client oldClient = (Client) oldClientRaw;
+                        if (oldClientRaw != null && !oldClient.getApiKey().equals(client.getApiKey())) {
+                            objectMap.remove(oldClient.getApiKey(), endFuture.completer());
+                        } else {
+                            endFuture.complete();
+                        }
+                    }, endFuture)
+                    // When finished, call this handler and then resultHandler
+                    .setHandler(handleResult(resultHandler));
+                }
+            } else {
+                resultHandler.handle(AsyncResultImpl.create(compositeResult.cause()));
+            }
+        });
+    }
+
+    @Override
+    public void unregisterClient(Client client, IAsyncResultHandler<Void> resultHandler) {
+        String clientIndex = getClientIndex(client);
+        objectMap.get(clientIndex, handleSuccessfulResult(resultHandler, oldClientRaw -> {
+            Client oldClient = (Client) oldClientRaw;
+            if (oldClient == null) {
+                Exception ex = new RegistrationException(Messages.i18n.format("InMemoryRegistry.ClientNotFound"));
+                resultHandler.handle(AsyncResultImpl.create(ex));
+            } else {
+                Future<Object> future1 = Future.future();
+                Future<Object> future2 = Future.future();
+
+                objectMap.remove(clientIndex, future1.completer());
+                objectMap.remove(oldClient.getApiKey(), future2.completer());
```

Comment:
This is from your design, I think.

There are references that go e.g. `api-key-123 -> client` and ` CLIENT::ORG|CLIENT-ID|VERSION-> client`. 

I had to make things a bit more circuitous than I'd like because there are no transactions.

Was that what were asking?